### ISSUE TITLE
Trim RFH header

### ISSUE THREAD

**Issue by Pzixel** [dsuch/pymqi#96]
Related to #44 

Currently instead of useful XML I get trashed message such as

>u'RFH \x02\x00\x00\x00L\x02\x00\x00"\x02\x00\x00\ufffd\x04\x00\x00MQSTR   \x00\x00\x00\x00\ufffd\x04\x00\x00\ufffd\x00\x00\x00<jms><Dst>MPGU.APPLICATION</Dst><Dlv>2</Dlv><Exp>0</Exp><Pri>4</Pri><Tms>1552293890271</Tms><Cid> </Cid><Rto> </Rto><Type>jms_bytes</Type></jms>L\x01\x00\x00<usr><ApplicationId>159954223</ApplicationId><DeclarantType>LE</DeclarantType><FormId>030301</FormId><OrgId>dtiu</OrgId><SSOId>6efe6f76-1c0b-4f3e-ad04-41dafe8453a1</SSOId><ServerAddress>10.89.84.192</ServerAddress><ServerName>g.local</ServerName><ServiceNumber>0001-9000003-030301-0359090/19</ServiceNumber></usr> @\x00\x00\x00<mcd><Msd>xmlnsc</Msd><Set></Set><Type></Type><Fmt></Fmt></mcd> <?xml version="1.0" encoding="UTF-8" standalone="yes"?><v5:StatusMessage xmlns:v5="http://asguf.xx/rkis_gu/coordinate/v5/"><v5:ResponseDate>2019-03-11T11:49:57.484+03:00</v5:ResponseDate><v5:PlanDate xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:nil="true"/><v5:StatusCode>77061</v5:StatusCode><v5:Documents><v5:ServiceDocument><v5:DocCode>18</v5:DocCode><v5:DocSubType>1</v5:DocSubType><v5:DocNumber>\u0431/\u043d</v5:DocNumber><v5:DocDate xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:nil="true"/><v5:ValidityPeriod xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:nil="true"/><v5:ListCount xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:nil="true"/><v5:CopyCount xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:nil="true"/><v5:DocFiles/></v5:ServiceDocument></v5:Documents><v5:Contacts/><v5:ServiceNumber>0001-9000003-030301-0359090/19</v5:ServiceNumber></v5:StatusMessage>'

When I want

> <xml version="1.0" encoding="UTF-8" standalone="yes"?><v5:StatusMessage xmlns:v5="http://asguf.xx/rkis_gu/coordinate/v5/"><v5:ResponseDate>2019-03-11T11:49:57.484+03:00</v5:ResponseDate><v5:PlanDate xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:nil="true"/><v5:StatusCode>77061</v5:StatusCode><v5:Documents><v5:ServiceDocument><v5:DocCode>18</v5:DocCode><v5:DocSubType>1</v5:DocSubType><v5:DocNumber>\u0431/\u043d</v5:DocNumber><v5:DocDate xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:nil="true"/><v5:ValidityPeriod xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:nil="true"/><v5:ListCount xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:nil="true"/><v5:CopyCount xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:nil="true"/><v5:DocFiles/></v5:ServiceDocument></v5:Documents><v5:Contacts/><v5:ServiceNumber>0001-9000003-030301-0359090/19</v5:ServiceNumber></v5:StatusMessage>

Currently I'm just trimming message:

```
raw_message = queue.get(None, md, options)
message_start_index = raw_message.index('<?xml')
message = raw_message[message_start_index:]
```
But it doesn't work very well since not all XML start with `<?xml`, and it looks very hacky.

**Comment by dsuch** [comment]
@Pzixel - please watch your language. This is not a support forum for wants expressed in an impolite manner.

I probably understand what your needs are but there will be no discussion on this level. 

I am going to close this ticket - do not come back until you start to appreciate it that if you expect free labour from strangers on the Internet then you need to be respectful and tactful towards them.

**Comment by dsuch** [comment]
@Pzixel The matter is that MQRFH2 is not a documented format and while it is possible to understand it to some extent, I am somewhat reluctant to add any complex support for it until IBM document it.

It is known that, for instance, where in this message, at which byte, information about the beginning of the XML content is.

However, JMS is but a part of what MQRFH2 can do and I am concerned that when we add it then the next, quite logical, expectation will be to support arbitrary headers produced by IBM ACE (Message Broker) and this is something that we cannot do because the format/protocols are not documented.

I think at this point there is no really better way than to scan the string to find the XML message marker. 

Or, even if instead of scanning we first parse MQRFH2 to extract the byte that points to XML and only then we return XML itself, we will will not be able to fully parse MQRFH2 and get all of its headers - again, simply because there is no information what it may possibly contain.

To sum up - we could somehow get the XML but parsing of MQRFH2 is something I would rather not add at this point

**Comment by Pzixel** [comment]
I see.

Anyway, thanks for your time.

**Comment by JoshMcMQ** [comment]
Hi @dsuch, RFH2 was IBM's way of passing the JMS headers across MQ along with user defined message properties.  The JMS headers are almost all a similar to already existing MQMD headers, ex: MessageID vs JMSMessageID.  In fact if a JMS app reads a message that has no RFH2 it will populate the JMS header values based on the MQMD values.

Up through v6 this RFH2 header was just part of the message body and apps that were not JMS would need to deal with it or instruct the sending JMS app to to suppress the RFH2 header by speifying the target client was MQ not JMS .

In v7.0 and higher the queue manager now processes the RFH2 header like any other header (under the covers MQ it was described in various placed that MQ consumes the RFH2 header during the PUT call and then regenerates it during the GET call if required), and all APIs were extended to have user defined message properties in the MQMD.  For backward compatibility the queue manager by default  [queues will have PROPCTL(COMPAT)] will try to provide the message to the app as it would be expected on the version of MQ that is connecting, so for instance if a v6 C program was connecting to a v6 queue manager it would see the RFH2 header in the message body if one was present, the same would be true if a v6 C program was connecting to a v7.0 or higher queue manager.

At v7.0 and higher other get message options have been added such as `MQGMO_NO_PROPERTIES`, `MQGMO_PROPERTIES_FORCE_MQRFH2`, and `MQGMO_PROPERTIES_IN_HANDLE`  to allow you to force the behavior you want.  For example you could use `MQGMO_PROPERTIES_FORCE_MQRFH2` if you had a program that expected the RFH2 in the message body but after upgrade to v7.0 it was no longer found.

If the goal is only to suppress the RFH2 headers then just add `MQGMO_NO_PROPERTIES`.

If the goal is to read the values of the name value pairs (both user defined and native RFH2 values) then add `MQGMO_PROPERTIES_IN_HANDLE`, this still gets rid of the RFH2 output from the message body but lets you inquire the properties via `MQINQP`.  See the `printProperties` function in the `/opt/mqm/samp/amqsbcg0.c` for an example of how this can be accomplished.  Maybe pymqi could use a similar c function to provide a dictionary of the message properties?

**Comment by Pzixel** [comment]
@dsuch 

> Would that work for you? I ask because I originally thought you were only interested in XML but now I am not sure - would you like to have access to MQRFH2 data as well?

No, it won't, because sometime XML arrives without XML headers and possible with/without namespace. So I actually ended with:

```py
raw_message = queue.get(None, md, options)
message_start_index = re.search("<(\w+:)?((StatusMessage)|(Message))", raw_message).start()
message = raw_message[message_start_index:]
```

Where I know that root tag is either `StatusMessage` or `Message`. It won't work in general case of course.

**Comment by JoshMcMQ** [comment]
@Pzixel I think you just need to do this:

```
options = pymqi.GMO()
options.Options = pymqi.CMQC.MQGMO_NO_PROPERTIES | pymqi.CMQC.MQGMO_FAIL_IF_QUIESCING
raw_message = queue.get(None, md, options)
```

Try it and let us know.

---

@dsuch, I noted pymqi already has a `_Properties` class, this has a method get and getitem that allow you to find a specific item.  What it lacks is something to return a dict of all properties, this is where the amqsbcg `printProperties` function can provide inspiration on looping through all properties on the message.

**Comment by dsuch** [comment]
Hello,

I went ahead to test it, which worked as expected, and to release a new version with an additional [**.get_no_jms**](https://dsuch.github.io/pymqi/examples.html#how-to-get-a-message-without-jms-mqrfh2-headers) method to support this use-case.

Please install v. 1.9.2 and confirm it, @Pzixel.

**Comment by JoshMcMQ** [comment]
@dsuch in the example, the second example is same as first, I think it should be `get_no_rfh2`?

> ```
> # Get the message but discard any JMS headers
> message = queue.get_no_jms()
> 
> # Works exactly as above: get_no_rfh2 is an alias to get_no_jms
> message = queue.get_no_jms()
> ```

**Comment by dsuch** [comment]
Thanks @JoshMcMQ - this is corrected now.

**Comment by Pzixel** [comment]
I was using `pymqi.CMQC.MQGMO_WAIT | pymqi.CMQC.MQGMO_NO_SYNCPOINT;` so I added `MQGMO_NO_PROPERTIES` and it worked. Thanks for help guys.

**Comment by dsuch** [comment]
Thanks @Pzixel and @JoshMcMQ - I am glad that it works and that we have a new feature along with a usage example!

**Comment by JoshMcMQ** [comment]
@Pzixel Note that @dsuch added a new call `queue.get_no_jms` which will automatically add the `MQGMO_NO_PROPERTIES` for you.

**Comment by Pzixel** [comment]
Yep, but I already use other options so it may be useful to have all of them in one place.

**Comment by JoshMcMQ** [comment]
@Pzixel if I read his update right it will just add `MQGMO_NO_PROPERTIES` to any options that you pass in to the call.  If you pass no options it will default to `MQGMO_NO_PROPERTIES` 
 with `MQGMO_FAIL_IF_QUIESCING`.
